sklearn/metrics/tests/test_classification.py

```diff
@@ -3395,3 +3399,202 @@ def test_d2_log_loss_score_raises():
     err = "The labels array needs to contain at least two"
     with pytest.raises(ValueError, match=err):
         d2_log_loss_score(y_true, y_pred, labels=labels)
+
+
+def test_d2_brier_score():
+    sample_weight = [2, 2, 3, 1, 1, 1]
+    # Binary targets
+    y_true = [0, 1, 1, 0, 0, 1]
+    y_true_string = ["no", "yes", "yes", "no", "no", "yes"]
+
+    # check that the value of the returned d2 score is correct
+    y_proba = [0.3, 0.5, 0.6, 0.7, 0.9, 0.8]
+    y_proba_ref = [0.5, 0.5, 0.5, 0.5, 0.5, 0.5]
+    d2_score = d2_brier_score(y_true=y_true, y_proba=y_proba)
+    brier_score_model = brier_score_loss(y_true=y_true, y_proba=y_proba)
+    brier_score_ref = brier_score_loss(y_true=y_true, y_proba=y_proba_ref)
+    d2_score_expected = 1 - brier_score_model / brier_score_ref
+    assert pytest.approx(d2_score) == d2_score_expected
+
+    # check that a model which gives a constant prediction equal to the
+    # proportion of the positive class should get a d2 score of 0
+    y_proba = [0.5, 0.5, 0.5, 0.5, 0.5, 0.5]
+    d2_score = d2_brier_score(y_true=y_true, y_proba=y_proba)
+    assert d2_score == 0
+    d2_score = d2_brier_score(y_true=y_true_string, y_proba=y_proba, pos_label="yes")
+    assert d2_score == 0
+
+    # check that a model which gives a constant prediction equal to the
+    # proportion of the positive class should get a d2 score of 0
+    # when we also provide sample weight
+    y_proba = [0.6, 0.6, 0.6, 0.6, 0.6, 0.6]
+    d2_score = d2_brier_score(
+        y_true=y_true, y_proba=y_proba, sample_weight=sample_weight
+    )
+    assert d2_score == 0
+    d2_score = d2_brier_score(
+        y_true=y_true_string,
+        y_proba=y_proba,
+        sample_weight=sample_weight,
+        pos_label="yes",
+    )
+    assert d2_score == 0
+
+    # Multiclass targets
+    y_true = [0, 2, 1, 3, 2, 0]
+    y_true_string = ["aa", "cc", "bb", "dd", "cc", "aa"]
+
+    # check that a model which gives a constant prediction equal to the
+    # proportion of the given labels gives a d2 score of 0 when we also
+    # provide sample weight
+    y_proba = [
+        [0.3, 0.3, 0.3, 0.1],
+        [0.3, 0.3, 0.3, 0.1],
+        [0.3, 0.3, 0.3, 0.1],
+        [0.3, 0.3, 0.3, 0.1],
+        [0.3, 0.3, 0.3, 0.1],
+        [0.3, 0.3, 0.3, 0.1],
+    ]
```

Comment:
I would prefer some non-symmetric data, like
```python
y_true = [0, 1, 1, 1, 2, 2, 2, 2, 3, 3]
y_true_string = ["dd", "cc", "cc", "bb", "bb", "bb", "bb, "aa"]
y_proba = np.tile([
        [0.1, 0.3, 0.4, 0.2],  # not 100% sure about the correct order
    , ...)
```

In the current state, passing label or not has little test effectiveness due to too much symmetry.